### ISSUE TITLE
Display settings

### ISSUE THREAD

**Issue by ivan-ristovic** [matf-pp/2023_Pesak#13]
Is there a way to set the display settings (e.g., the window resolution and FPS)?

If the codebase allows it, I would consider adding a "toggle" button between 30/60/144/uncapped FPS and a choice of a few resolution values (both 4:3 and 16:9 if possible).

**Comment by Spera97** [comment]
о томе смо размишљали, стоји нам на неком тудуу

**Comment by ivan-ristovic** [comment]
There wasn't a ticket for this nor a [kanban](https://github.com/matf-pp/2023_Pesak/projects?query=is%3Aopen) so let's discuss the progress in the comments as development goes on.

**Comment by sajberk** [comment]
The option to choose the two values is still only present in the source code. A lot of the physics directly depends on the game framerate and the resolution of the simulation, so playing at 144 FPS feels much faster than playing at 30 FPS. Because of that, we've had the idea to use the FPS limiting to control the speed of the simulation rather than it's "smoothness". We would take 60 FPS as normal speed, so 30 and 120 FPS would give us a simulation at half and double the normal speed, respectively. 

Assuming square pixels, the window resolution has to be a multiple of the resolution of the underlying physics simulation. Currently, we are calculating the scaling factor that would cover around 80% of the player's screen and scaling everything on the screen to fit that, so the scaling mechanism is already present in the code and working. Because of that, our idea is to use it as the default, and eventually add two buttons that would lower and raise it, as well as another button that would change the aspect ratio. As there would usually be very few possible resolutions (excluding very small windows), a single button that cycles through all the combinations would also be possible, and would probably make for a cleaner interface. Do any of these sound like good options?

The simulation resolution itself needs to be a fixed value, as all of the game physics depends on it. We would like to choose a simulation size that even lower-end hardware could simulate at 120 frames per second, but which also looks natural on the default 60 FPS speed. We are still trying to squeeze out every possible FPS out of our code, so this will be decided in the coming days, as we work on the few optimizations that we have in mind.

In the end, I just want to give you a quick heads up that you've probably tried our game with the simulation size i used for testing a new feature yesterday after the "stable" release, as that was the latest commit when you opened this issue. I forgot to change it back to the current default value before pushing. (a gitignored config file would have helped, i know 😅) We never intended for the game to run in a tiny square, so if you've played it like that, you might want to revisit it with the proper resolution.

**Comment by ivan-ristovic** [comment]
It all sounds good, I don't mind the switches being a single button, one for FPS and one for the resolution and/or aspect ratio.

I am curious how CPU intensive is the simulation? In other words, what is the max FPS you can achieve with the current grid size and what is the max grid size you can have with 60 FPS?

**Comment by sajberk** [comment]
Well, until a few days ago we often joked about repurposing Pesak as benchmarking software, since it used to completely hammer the CPU. Now it runs much better (since we switched from software-rendering tens of thousands of rectangles to hardware-rendering a single texture), and CPU usage now usually sits at 10-30% depending on the CPU. We didn't want to put much time into optimizing the game before the deadline, and opted for adding more features and fixing known bugs instead. As for the optimizations, the first one will definitely be replacing hash maps with runtime generated arrays, since hash map access is currently taking *a third* of Pesak's total CPU time. We are also thinking about possibly adding double-buffering as well as some level of multithreading.

As for the exact numbers, we liked the 240x144 resolution for testing, and it gives us around 100 FPS with a screen full of particles. It's worth noting that everyone on our team has relatively dated hardware (I'm using an i7-6700HQ), and some of our friends with recent CPUs are seeing almost double the numbers. 
On my machine, I can't go over around 60k pixels total with stable 60 FPS, but we are hoping to see the numbers go up a lot as we proceed with the planned optimizations.
